validator/src/journal/chain_ffi.rs

```diff
@@ -0,0 +1,723 @@
+/*
+ * Copyright 2018 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ------------------------------------------------------------------------------
+ */
+use cpython;
+use cpython::{FromPyObject, ObjectProtocol, PyList, PyObject, Python, PythonObject, ToPyObject};
+use journal::block_wrapper::BlockWrapper;
+use journal::chain::*;
+use py_ffi;
+use pylogger;
+use std::ffi::CStr;
+use std::os::raw::{c_char, c_void};
+use std::sync::mpsc::Sender;
+use std::thread;
+
+use batch::Batch;
+
+use protobuf::Message;
+
+use proto::transaction_receipt::TransactionReceipt;
+
+#[repr(u32)]
+#[derive(Debug)]
+pub enum ErrorCode {
+    Success = 0,
+    NullPointerProvided = 0x01,
+    InvalidDataDir = 0x02,
+    InvalidPythonObject = 0x03,
+    InvalidBlockId = 0x04,
+
+    Unknown = 0xff,
+}
+
+#[no_mangle]
+pub extern "C" fn chain_controller_new(
+    block_store: *mut py_ffi::PyObject,
+    block_cache: *mut py_ffi::PyObject,
+    block_validator: *mut py_ffi::PyObject,
+    chain_head_lock: *mut py_ffi::PyObject,
+    on_chain_updated: *mut py_ffi::PyObject,
+    observers: *mut py_ffi::PyObject,
+    data_directory: *const c_char,
+    chain_controller_ptr: *mut *const c_void,
+) -> ErrorCode {
+    if block_store.is_null() {
+        return ErrorCode::NullPointerProvided;
+    }
```

Comment:
I wrote a little macro to do this in the publisher ffi code that you could reuse:
```
macro_rules! check_null {
    ($($arg:expr) , *) => {
        $(if $arg.is_null() { return ErrorCode::NullPointerProvided; })*
    }
}
```

Reply:
https://github.com/boydjohnson/sawtooth-core/blob/block.publisher.rust/validator/src/journal/publisher_ffi.rs#L61

Reply:
I will add this.